### DISCUSSION TITLE
Seeking best practice advice on using Measure tools in reality model

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2267] [Q&A]
A user has sent the following issue related to using measure tools in reality models.  He sent the following screen caps of his issue (please ignore his comments in red).  It shows 3 points on reality model (left image) he selected for measurement (using the measure tool).  The points appear to be at locations where he wants to measure.  When the model is rotated slightly (right image), 2 points are in completely different location where he intended to measure.

I realize this is probably because of converting 2d screen coordinates to 3d world coordinate, I would like to get guidance on what would be the best way for the use to make sure he can selected the intended points.  Perhaps there is a snap setting where this works better.

This was the reason for my earlier post #2060, with hopes that it will give user better visual guidance on picking points.  Until this capability is available, if there is recommendation I can provide to user such that it helps him pick locations on his reality model more accurately, it will help him with his workflows - measurement seems to his most common use.

Thank you,

![Measure Error - Reality Model](https://user-images.githubusercontent.com/83007711/133594094-d43021c3-4e59-4003-a317-c93f1fbb6689.jpg)

**Comment by pmconne** [comment]
I fail to reproduce this problem using the Measure Location and Measure Distance tools in iTwin Design Review - the tools correctly identify the location on the reality mesh and display the marker properly.
Measure...
![image](https://user-images.githubusercontent.com/22944042/133607771-9a530083-9eb3-4d18-95b2-593cdf8facbf.png)
...then rotate.
![image](https://user-images.githubusercontent.com/22944042/133607991-43406ea9-897e-47e4-9791-cf5266044643.png)

However, I gather that the Design Review tools are not the same as those in the imodeljs-frontend package, which I assume are the ones you are using @ovaisa?
@bbastings do those tools suffer this problem? I am currently unable to test them.

**Reply by aruniverse** [reply]
The meausre-tools used in Design Review are available for 3rd parties to use with the [@bentley/measure-tools-react](https://www.npmjs.com/package/@bentley/measure-tools-react) package.

**Reply by bbastings** [reply]
@pmconne I am neither aware nor can I repro issues with using the imodeljs-frontend measure tools and snapping to the reality mesh. Some things that can affect the snap point location are AccuDraw, ACS Plane Snap, or having the "Locate" option disable for terrain (points will be projected to view's ACS plane)...but those would need to be explicitly set. Suggest changing snap mode to nearest so you can see the "yellow x" snap icon tracking the surface.

**Reply by pmconne** [reply]
@ovaisa is your app using the MeasureDistanceTool and MeasureLocationTool delivered with imodeljs-frontend?
Does it make any changes to the settings @bbastings mentioned?

**Reply by ovaisa** [reply]
Thank you all for your response to my question.  

I am not using default imodeljs-frontend toolbars, I defined my own but using the Core measure tools by using the following call when creating toolbars.  This tool group contains both MeasureDistanceTool and MeasureLocationTool

      ToolbarHelper.createToolbarItemFromItemDef(1000, CoreTools.measureToolGroup),

I will look into @bbastings suggestion by turning on those options and see if it improves users experience.  Thanks for the link to measure tools used by Design Review.

**Reply by bbastings** [reply]
@ovaisa The only option you want to ensure is on is locate for the reality model; ACS Plane Snap should remain off...I only mentioned it and AccuDraw as they can adjust the snap point.
